West docs for 2.5

mbolivar-nordic opened the pull request:
Miscellaneous fixes and improvements from reviewing the documentation.

mbolivar-nordic committed:
doc: sort redirects

Just to keep things easier to find.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: clean up west index

Rename the 'basics' page to basics.rst, adding a redirect for the old
title.

Rearrange some other index pages to improve the flow:

- Put relatively stale content in moving-to-west.rst to the bottom. I
  expect pepole have either moved west or are choosing not to.

- Promote the release notes page higher up, since that seems more
  important.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: glossary: update west terms

We've got some obsolete 'west installation' usages kicking around
still.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: west: move advanced content out of 'basics'

With the documentation for west 0.9, the 'basics' page is completely
out of control. Reorganize it so it fits in a few pages, moving many
details out into other pages for the truly curious.

As part of this, create new pages for built-in commands (built-in.rst)
and esoteric details about workspaces (workspaces.rst).

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: update west's workspaces.rst for 2.5

We haven't updated the example revisions in the alternate topologies
sections in a while. Update these for the upcoming 2.5 release.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: add missing docs for 'west zephyr-export'

Something should have been added to zephyr-cmds.rst when this
extension was added.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: improve manifest key docs

Create tables and Sphinx sections for the sub-keys of a manifest.

List the subsections in the same order they are documented.

This makes it easier to jump around in the document from the table of
contents and see what's in each subkey.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: improve docs on building without west

Make it more obvious what needs to happen by spelling out the
commands.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

tejlmand commented on doc/conf.py:
```diff
@@ -343,28 +330,28 @@
     ('boards/arm/nrf51_blenano/doc/nrf51_blenano', 'boards/arm/nrf51_blenano/doc/index'),
-    ('boards/arm/nrf51_pca10028/doc/nrf51_pca10028', 'boards/arm/nrf51dk_nrf51422/doc/index'),
     ('boards/arm/nrf51_pca10028/doc/index', 'boards/arm/nrf51dk_nrf51422/doc/index'),
+    ('boards/arm/nrf51_pca10028/doc/nrf51_pca10028', 'boards/arm/nrf51dk_nrf51422/doc/index'),
```
Just curious.
The pca board were deprecated before 2.3, and completely removed with this PR: https://github.com/zephyrproject-rtos/zephyr/pull/28763

How long do we want to keep the redirection link for the old boards ?

mbolivar-nordic replied:
> How long do we want to keep the redirection link for the old boards ?

It's a fair question, but it's out of scope for this PR.

tejlmand commented on doc/guides/west/basics.rst:
```diff
@@ -1,19 +1,16 @@
+repositories) under a common *workspace* directory.
+
+Example workspace
+*****************
```
You do some very nice cleanups in this PR, so maybe you could cleanup this as well.
Comment belong to those lines that are not touched in the PR: https://github.com/zephyrproject-rtos/zephyr/blob/38bac6da34ac47e6e79e919e617a0ce9b8367310/doc/guides/west/basics.rst#L18-L30

At least I find it this hard to read.

Maybe something like:
```
zephyrproject/                 # west topdir
├── .west/                     # marks the location of the topdir
│   └── config                 # per-workspace local configuration file
│
│   # Below, the manifest repository, never modified by west after creation
├── zephyr/                    # .git/ repo
│   ├── west.yml               # manifest file
│   └── [... other files ...]
│
│   # Below, projects updated by west
├── modules/
│   └── lib/
│       └── tinycbor/          # .git/ project
├── net-tools/                 # .git/ project
└── [ ... other projects ...]
```

mbolivar-nordic replied:
> At least I find this hard to read.

Good idea. I added another patch.

tejlmand reviewed (commented):
Looking good.
Just two minor comments.

mbolivar-nordic committed:
doc: west: clean up example workspace

Suggested-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

marc-hb commented on doc/guides/west/workspaces.rst:
```diff
@@ -0,0 +1,309 @@
+
+- T1: star topology, zephyr is the manifest repository
+- T2: star topology, a Zephyr application is the manifest repository
+- T3: forest topology, freestanding manifest repository
```
I finally took the time to understand why these analogies repeatedly failed to help me remember their corresponding organizations.

The main issue for me is that a "topology" is normally something spatial, so I kept looking for _directory_ structures looking like "stars" or "forests". Fail. The immediately previous section is all about directories, sure didn't help.

Once lost there it was very hard for me to see that the "star" is a git _repo_ (barely related to directories) and that what makes this repo "the star" is owning the manifest. Obvious... with hindsight.

It's probably too late now to replace "topology" by something less "spatial" ("organization"?) but there could be at least a (non-)transition sentence that clearly departs this section from the previous, directory-focused section and that defines a bit what "topology" means here. For instance:

```
The following are example source code topologies supported by west. A "topology" refers to which git repository is the manifest versioned in.
- T1: star topology, zephyr is the "star" repository because it holds the manifest
- T2: star topology, a Zephyr application is the manifest repository
```

mbolivar-nordic replied:
I would like to defer improvements to this section until after 2.5. There is some great parallel work in this arena (https://github.com/zephyrproject-rtos/example-application/) that I want to see play out before attempting to bikeshed this.

In no way am I trying to say it's not worth fixing, though. I find this section confusing too.

marc-hb commented on doc/guides/west/workspaces.rst:
```diff
@@ -0,0 +1,309 @@
+you manually add to this branch may be lost the next time you run ``west
+update``. Instead, check out a local branch with another name, and either
+rebase it on top of a new ``manifest-rev``, or merge ``manifest-rev`` into
+it.
```
... Some people find useful to make their branches track `manifest-rev`:
```
git branch --set-upstream-to=manifest-rev
git status
Your branch and 'manifest-rev' have diverged,
and have 1 and 2 different commits each, respectively.
```
```

mbolivar-nordic replied:
> ... Some people find useful to make their branches track `manifest-rev`:

Maybe we can make a 'tips and tricks' page sometime, but I'm not going to add this recommendation in this PR.

marc-hb commented on doc/guides/west/troubleshooting.rst:
```diff
@@ -5,6 +5,36 @@ Troubleshooting West
+The ``git fetch`` command example in the last line above is what needs to
+succeed. Go to ``/some/directory``, copy/paste and run the entire ``git fetch``
+command, then debug from there using the documentation for your credential
+storage helper.
```
git has most often proved useless to debug proxy, authentication or other network issues. It's generally required to peel yet another layer of indirection and use `curl -v` or `ssh -v`. Maybe mention that briefly.

mbolivar-nordic replied:
> `curl -v` or `ssh -v`. Maybe mention that briefly.

Easy enough and sounds wise. Added a commit.

marc-hb commented on doc/guides/west/manifest.rst:
```diff
@@ -181,55 +201,87 @@ In this manifest:
+   * - ``clone-depth``
+     - Optional. If given, a positive integer which creates a shallow history
+       in the cloned repository limited to the given number of commits. This
+       can only be used if the ``revision`` is a branch or tag.
```
Curious how this will interact with the incoming `shallow config`...

mbolivar-nordic replied:
> Curious how this will interact with the incoming `shallow config`...

This is basically a failed experiment that no one uses, so it's a bit of a tree falling in the forst.

marc-hb reviewed (commented):
OK, I only now realize this is a backport. I was wondering why most of this was getting added again... sorry for posting these comments in the wrong place - not sure what a better place would be.

mbolivar-nordic commented:
Hm? It's not a backport, but one commit does move a lot of stuff around. Thanks for review.

mbolivar-nordic committed:
doc: west: improve troubleshooting for west update

Recommend curl -v or ssh -v.

Suggested-by: Marc Herbert <marc.herbert@intel.com>
Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

marc-hb reviewed (commented):
> It's not a backport, but one commit does move a lot of stuff around.

I missed the large deletion, github collapses it by default. Never mind.

carlescufi reviewed (approved):
All commits look good to me.
